pkg/querier/worker/scheduler_processor_test.go

```diff
@@ -97,10 +100,63 @@ func TestSchedulerProcessor_processQueriesOnSingleStream(t *testing.T) {
 		// and then to send the query result.
 		loopClient.AssertNumberOfCalls(t, "Send", 2)
 		loopClient.AssertCalled(t, "Send", &schedulerpb.QuerierToScheduler{QuerierID: "test-querier-id"})
+
+		require.Equal(t, 1, int(frontend.queryResultCalls.Load()), "expected frontend to be informed of query result exactly once")
+	})
+
+	t.Run("should report query result to frontend even if query is cancelled", func(t *testing.T) {
+		sp, loopClient, requestHandler, frontend := prepareSchedulerProcessor(t)
+
+		recvCount := atomic.NewInt64(0)
+		queryEvaluationBegun := make(chan struct{})
+
+		loopClient.On("Recv").Return(func() (*schedulerpb.SchedulerToQuerier, error) {
+			switch recvCount.Inc() {
+			case 1:
+				return &schedulerpb.SchedulerToQuerier{
+					QueryID:         1,
+					HttpRequest:     nil,
+					FrontendAddress: frontend.addr,
+					UserID:          "user-1",
+				}, nil
+			default:
+				// Wait until query execution has begun, then simulate the scheduler shutting down.
+				<-queryEvaluationBegun
+				return nil, schedulerpb.ErrSchedulerIsNotRunning
+			}
+		})
+
+		requestHandler.On("Handle", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+			ctx := args.Get(0).(context.Context)
+
+			// Trigger the shutdown of the scheduler.
+			close(queryEvaluationBegun)
+
+			// Wait for our context to be cancelled before continuing.
+			select {
+			case <-ctx.Done():
+				// Nothing more to do.
+			case <-time.After(time.Second):
+				require.Fail(t, "expected query execution context to be cancelled when scheduler shut down")
+			}
+		}).Return(&httpgrpc.HTTPResponse{}, nil)
+
+		workerCtx, workerCancel := context.WithCancel(context.Background())
+
+		// processQueriesOnSingleStream() blocks and retries until its context is cancelled, so run it in the background.
+		go func() {
+			sp.processQueriesOnSingleStream(workerCtx, nil, "127.0.0.1")
+		}()
+
+		require.Eventually(t, func() bool {
+			return frontend.queryResultCalls.Load() == 1
+		}, time.Second, 10*time.Millisecond, "expected frontend to be informed of query result exactly once")
```

Comment:
(nit) Will this actually check if it is exactly once? It's possible that tick happens inbetween a result being returned where the conditional function returns true already.

Perhaps another assert after workCancel below to ensure it was only once?

Reply:
The test sets up the fake scheduler to only ever return one query, so only one query result should ever be returned to the frontend. (Without the fix in this PR, the frontend received no results at all, so observing one result being returned indicates the issue has been fixed.)

If the querier was calling the frontend multiple times for some reason, then I'd expect this test to flake occasionally.